content/specifications/v0.1.0.md

```diff
@@ -467,4 +467,182 @@ user_preferences::
+  ;
+
+multiline_list:
+  (* preceded(NEWLINE, ...) will error on hanging newline at end *)
+  | multiline_list_item, multiline_list_items
+  ;
+
+multiline_list_items:
+  | NEWLINE, multiline_list_item, multiline_list_items
+  | NEWLINE?
+  ;
+
+multiline_list_item:
+  | DASH ( scalar | vector_value )
+  ;
+
+(* dicts *)
+
+inline_dict:
+  | EMPTY_DICT
+  | scalar_key_value, ( COMMA scalar_key_value )*
+  ;
+
+multiline_dict:
+  | multiline_dict_item, multiline_dict_items
+
+multiline_dict_items:
+  | NEWLINE, multiline_dict_item, multiline_dict_items
+  | NEWLINE?
+  ;
+
+multiline_dict_item:
+  | scalar_key_value
+  | dict_key, vector_value
+  ;
+
+scalar_key_value:
+  | dict_key, SCALAR_START, scalar
+  ;
+
+dict_key:
+  | IDENT | STRING ;
+
+vector_value:
+  | INLINE_VECTOR_START, inline_vector
+  | MULTILINE_VECTOR_START, NEWLINE, INDENT, multiline_vector, DEDENT
+  ;
+```
+
+## Tokenizer
+
+### Tokens
+- NEWLINE
+- EOF
+- COMMA
+- INDENT
+- DEDENT
+- <string> IDENT
+- <string> STRING
+- <float> FLOAT
+- <int> INT
+- <bool> BOOL
+- NULL
+- EMPTY_LIST
+- EMPTY_DICT
+- SCALAR_START
+- INLINE_VECTOR_START
+- MULTILINE_VECTOR_START
+
+### Production rules
+
+`curr_ind` = current indentation level (in spaces).
+`curr_ind += 2` at each `INDENT` and `curr_ind -= 2` at each `DEDENT`.
+
+```
+| '\n', whitespace*, comment              -> skip & continue
+| '\n', ' '{curr_ind}                     -> NEWLINE
+| '\n', ' '{curr_ind+2}                   -> NEWLINE, INDENT
+| '\n', ' '{< curr_ind}                   -> DEDENT (curr_ind - len(m)) / 2 times, NEWLINE
+| (whitespace*, comment?, '\n')+          -> NEWLINE  # compress newlines together
+| "[]"                                    -> EMPTY_LIST
+| "{}"                                    -> EMPTY_DICT
+| "::", (whitespace*, comment)?, '\n'     -> [MULTILINE_VECTOR_START, NEWLINE]
+| "::"                                    -> INLINE_VECTOR_START  # expect space after
+| ":"                                     -> SCALAR_START         # expect space after
+| "-"                                     -> DASH                 # expect space after
+| ","                                     -> COMMA                # expect space after
+| ('+'|'-')?, ['0'-'9' '_']+              -> INT(int)             # from decimal
+| ('+'|'-')?, "0x",
+    ['0'-'9' 'a'-'f' 'A'-'F' '_']+        -> INT(int)             # from hex
+| ('+'|'-')?, "0o",
+    ['0'-'7' '_']+                        -> INT(int)             # from octal
+| ('+'|'-')?, "0b", ['0'-'1' '_']+        -> INT(int)             # from binary
+| "nan"                                   -> FLOAT(float)         # from nan
+| ('+'|'-')? "inf"                        -> FLOAT(float)         # from inf
+| int, '.', ['0'-'9' '_'],
+    (('e'|'E'), int)?                     -> FLOAT(float)
+| int, (('e'|'E') int)                    -> FLOAT(float)
+| "true" | "false"                        -> BOOL(bool)
+| "null"                                  -> NULL
```

Comment:
nuance missed here that `true` in the context of a dict must be interpreted as the dict key `"true"` e.g.
`true: 123` must be parsed as `{"true": 123}`. this also misses the case where a dict-key would start with `true` as the first 4 words.